core/iwasm/aot/arch/aot_reloc_riscv.c

```diff
@@ -170,14 +170,32 @@ rv_get_val(uint16 *addr)
     return ret;
 }
 
-/* Set a val to given address */
+/**
+ * Set a val to given address
+ *
+ * Zifencei extension is moved out from I extension in RISC-V spec 20191213
+ * (rv32i/rv64i version 2.1 and later), new version of GNU toolchain or clang
+ * may not support it if no explicit 'zifencei' is specified in arch attribute.
+ *
+ * Refer to:
+ * https://github.com/riscv-non-isa/riscv-c-api-doc/blob/master/riscv-c-api.md
+ */
+#if __riscv_i >= 2001000 && !defined(__ZEPHYR__)
+__attribute__((target("arch=+zifencei")))
+#endif
 static void
 rv_set_val(uint16 *addr, uint32 val)
 {
     *addr = (val & 0xffff);
     *(addr + 1) = (val >> 16);
 
+    /* REVISIT: Zephyr's GCC toolchain does not support attribute
+     * target("arch=+zifencei") so we must check if the extension is
+     * enabled at runtime.
+     */
+#if defined(__ZEPHYR__) && defined(CONFIG_RISCV_ISA_EXT_ZIFENCEI)
```

Comment:
Also note that GCC / Clang define the `__riscv_zifencei` symbol when supported: https://github.com/riscv-non-isa/riscv-c-api-doc/blob/master/riscv-c-api.md#architecture-extension-test-macro, which may be a more portable option here.